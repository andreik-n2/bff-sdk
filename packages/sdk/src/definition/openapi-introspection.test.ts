import path from 'path';
import { introspectOpenApiV2 } from './openapi-introspection';

describe('introspectOpenApiV2', () => {
	describe('introspection kind:string', function () {
		it('should introspect yaml string succesfully', async () => {
			const result = await (
				await introspectOpenApiV2({
					source: {
						kind: 'string',
						openAPISpec: `
openapi: 3.0.0
info:
  title: Example API
  version: 1.0.0

paths:
  /example:
    get:
      summary: Get example data
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                type: string`,
					},
				})
			)({});

			expect(result.DataSources[0].Custom.UpstreamSchema).toMatchSnapshot();
			expect(result.Schema).toMatchSnapshot('wg_schema');
		});

		it('should introspect json string succesfully', async () => {
			const result = await (
				await introspectOpenApiV2({
					source: {
						kind: 'string',
						openAPISpec: `
{
    "openapi": "3.0.0",
    "info": {
        "title": "Example API",
        "version": "1.0.0"
    },
    "paths": {
        "/example": {
            "get": {
                "summary": "Get example data",
                "responses": {
                    "200": {
                        "description": "OK",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "string"
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}`,
					},
				})
			)({});

			expect(result.DataSources[0].Custom.UpstreamSchema).toMatchSnapshot();
			expect(result.Schema).toMatchSnapshot('wg_schema');
		});
	});

	describe('introspection kind:object', function () {
		it('should introspect succesfully', async () => {
			const result = await (
				await introspectOpenApiV2({
					source: {
						kind: 'object',
						openAPIObject: {
							openapi: '3.0.0',
							info: {
								title: 'Example API',
								version: '1.0.0',
							},
							paths: {
								'/example': {
									get: {
										summary: 'Get example data',
										responses: {
											'200': {
												description: 'OK',
												content: {
													'application/json': {
														schema: {
															type: 'string',
														},
													},
												},
											},
										},
									},
								},
							},
						},
					},
				})
			)({});

			expect(result.DataSources[0].Custom.UpstreamSchema).toMatchSnapshot();
			expect(result.Schema).toMatchSnapshot('wg_schema');
		});
	});
});
