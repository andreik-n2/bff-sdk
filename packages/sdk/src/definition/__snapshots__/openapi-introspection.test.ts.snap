// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`introspectOpenApiV2 introspection kind:object should introspect succesfully 1`] = `
"schema {
  query: Query
}

directive @globalOptions(endpoint: String, operationHeaders: ObjMap, queryParams: ObjMap, queryStringOptions: ObjMap, sourceName: String) on OBJECT

directive @httpOperation(httpMethod: HTTPMethod, isBinary: Boolean, operationSpecificHeaders: ObjMap, path: String, queryParamArgMap: ObjMap, queryStringOptionsByParam: ObjMap, requestBaseBody: ObjMap) on FIELD_DEFINITION

enum HTTPMethod {
  CONNECT
  DELETE
  GET
  HEAD
  OPTIONS
  PATCH
  POST
  PUT
  TRACE
}

scalar ObjMap

type Query @globalOptions(sourceName: "api", operationHeaders: "{}") {
  "Get example data"
  example: String @httpOperation(path: "/example", operationSpecificHeaders: "{\\"accept\\":\\"application/json\\"}", httpMethod: GET)
}"
`;

exports[`introspectOpenApiV2 introspection kind:object should introspect succesfully: wg_schema 1`] = `
"schema {
  query: Query
}

enum HTTPMethod {
  CONNECT
  DELETE
  GET
  HEAD
  OPTIONS
  PATCH
  POST
  PUT
  TRACE
}

scalar ObjMap

type Query {
  "Get example data"
  example: String
}"
`;

exports[`introspectOpenApiV2 introspection kind:string should introspect json string succesfully 1`] = `
"schema {
  query: Query
}

directive @globalOptions(endpoint: String, operationHeaders: ObjMap, queryParams: ObjMap, queryStringOptions: ObjMap, sourceName: String) on OBJECT

directive @httpOperation(httpMethod: HTTPMethod, isBinary: Boolean, operationSpecificHeaders: ObjMap, path: String, queryParamArgMap: ObjMap, queryStringOptionsByParam: ObjMap, requestBaseBody: ObjMap) on FIELD_DEFINITION

enum HTTPMethod {
  CONNECT
  DELETE
  GET
  HEAD
  OPTIONS
  PATCH
  POST
  PUT
  TRACE
}

scalar ObjMap

type Query @globalOptions(sourceName: "api", operationHeaders: "{}") {
  "Get example data"
  example: String @httpOperation(path: "/example", operationSpecificHeaders: "{\\"accept\\":\\"application/json\\"}", httpMethod: GET)
}"
`;

exports[`introspectOpenApiV2 introspection kind:string should introspect json string succesfully: wg_schema 1`] = `
"schema {
  query: Query
}

enum HTTPMethod {
  CONNECT
  DELETE
  GET
  HEAD
  OPTIONS
  PATCH
  POST
  PUT
  TRACE
}

scalar ObjMap

type Query {
  "Get example data"
  example: String
}"
`;

exports[`introspectOpenApiV2 introspection kind:string should introspect yaml string succesfully 1`] = `
"schema {
  query: Query
}

directive @globalOptions(endpoint: String, operationHeaders: ObjMap, queryParams: ObjMap, queryStringOptions: ObjMap, sourceName: String) on OBJECT

directive @httpOperation(httpMethod: HTTPMethod, isBinary: Boolean, operationSpecificHeaders: ObjMap, path: String, queryParamArgMap: ObjMap, queryStringOptionsByParam: ObjMap, requestBaseBody: ObjMap) on FIELD_DEFINITION

enum HTTPMethod {
  CONNECT
  DELETE
  GET
  HEAD
  OPTIONS
  PATCH
  POST
  PUT
  TRACE
}

scalar ObjMap

type Query @globalOptions(sourceName: "api", operationHeaders: "{}") {
  "Get example data"
  example: String @httpOperation(path: "/example", operationSpecificHeaders: "{\\"accept\\":\\"application/json\\"}", httpMethod: GET)
}"
`;

exports[`introspectOpenApiV2 introspection kind:string should introspect yaml string succesfully: wg_schema 1`] = `
"schema {
  query: Query
}

enum HTTPMethod {
  CONNECT
  DELETE
  GET
  HEAD
  OPTIONS
  PATCH
  POST
  PUT
  TRACE
}

scalar ObjMap

type Query {
  "Get example data"
  example: String
}"
`;
